### DISCUSSION TITLE
Trouble with setting up Java modules in Apache NetBeans 20 Flatpak Installation, on Linux Mint 21.3 Virginia

### DISCUSSION THREAD

**Discussion by leandrocaplan** [apache/netbeans#7139] [Q&A] [answered]
Hi there!

I installed Apache NetBeans 20 using the Linux Mint MATE software manager. Initially, I created a simple "Hello world" project for testing purposes, and it worked fine.

However, when I attempted to open a Maven project obtained from a GitLab repository, NetBeans started downloading the Maven Central Repository Index from its official source ([https://repo.maven.apache.org/maven2/.i ... y-index.gz](https://repo.maven.apache.org/maven2/.index/nexus-maven-repository-index.gz)). Although the download took a reasonable amount of time, considering the size of the file (around 2 GB), it took two hours to extract its content. During this process, I noticed via htop that NetBeans was utilizing over 100% of CPU usage. Consequently, I stopped the process (never completed) and closed the IDE.

At that point, Maven and Tomcat were not installed on my system. I require both to properly run my projects. I attempted to uninstall NetBeans 20 using Flatpak and install NetBeans 21 from the .deb package downloaded from the official Apache NetBeans site. Unfortunately, the results were worse, as I couldn't even open a NetBeans project, as the IDE froze upon the attempt.

Subsequently, I reverted to the NetBeans 20 Flatpak installation and attempted to configure the "Maven Home" to my local Maven installation via the options window. However, NetBeans did not display "/etc/maven" or "/usr/share/maven" when browsing for the home directory. Also, it shows me an "/app" directory, which doesn't really exist in my filesystem. I'm unsure where NetBeans is obtaining this directory from. I think it may be a virtual one created by the IDE, or by Flatpak:
![netbeans](https://github.com/apache/netbeans/assets/90872467/17625826-520a-4523-b195-d17470ec9140)

Additionally, there is no "/.m2" folder in my home directory (and I understand that it should).

I can see those mentioned folders in my filesystem in the browser:
![maven_browser](https://github.com/apache/netbeans/assets/90872467/4429bc39-2468-434d-9a71-08a91bb6f8d3)

I don't know which one should be the Maven home directory, but NetBeans doesn't recognize me any of both.

I downloaded the repository index (.gz file) via my browser, but I'm unable to set up NetBeans to work with my native Maven installation. Furthermore, it does not recognize the OpenJDK 22 I installed via apt.

Any ideas on what might be happening? I feel like there are many things I am overlooking.

Thank you very much!

Leandro

**Comment by leandrocaplan** [comment, marked as answer]
I solved this issue myself by following these steps:

- Uninstalled both my Flatpak installation of Apache NetBeans 20 and my installation of NetBeans 21 from the ".deb" package.

- Installed Apache NetBeans 20 from a ".deb" package downloaded from the official webpage.

- In the terminal, ran the following command: "sudo update-alternatives --config java". Then, selected "/usr/lib/jvm/java-11-openjdk-amd64/bin/java" from the displayed options in the menu.

- Ran the command "java --version" in the terminal to verify if the previous step worked properly. Confirming that the default Java version in my system was indeed the one I had selected.

- Ran the command "whereis java" to see the directories where the desired JDK was installed.

The results in terminal looked like this:
[![console][2]][2]

- In the terminal, navigated to the root folder of the project and ran: 
```
sudo chown <my_username> -R .
sudo chmod 777 -R .
```

- Tested Maven by running the following command from that root folder to ensure it was working properly.
```mvn clean install -U -X```
- Edited the file "/etc/environment" with superuser privileges using the command "sudo xed /etc/environment". Added the following lines to the file:
```
JAVA_HOME="/usr/lib/jvm/java-11-openjdk-amd64"
export JAVA_HOME
```
- Edited the file "~/.bashrc" using the command "xed ~/.bashrc", and added these lines:
```
JAVA_HOME="/usr/lib/jvm/java-11-openjdk-amd64"
export PATH=$PATH:$JAVA_HOME/bin
export JAVA_HOME
```
- Edited the file "/usr/lib/apache-netbeans/etc/netbeans.conf" using the command "sudo xed /usr/lib/apache-netbeans/etc/netbeans.conf", and added the following line:
```netbeans_jdkhome=$JAVA_HOME```
- Rebooted the system, and then NetBeans ran with JDK 11 instead of JDK 21.

- Launched NetBeans, went to "Tools -> Options" in the menu bar, clicked the "Java" icon, and selected the "Maven" tab.

- Entered "/usr/share/maven" in the "Maven Home" field via the browser. Selected "JDK 11 (System)" in the "Default JDK" dropdown and entered "clean install -U -X" in the "Global Execution Options" field. Then, checked the platforms managed by the IDE by going to "Tools -> Java Platforms".

Those windows got like this:
[![netbeans_conf][1]][1]

-  Right-clicked on the root of the project in the left sidebar and selected "Properties" from the menu. In the properties window, navigated to "Build -> Compile" and selected "JDK 11 (System)" in the "Java Platform" field:
[![project_properties][3]][3]

Since I had Tomcat 9 installed on my system, I needed to stop and disable that service because my native installation of Tomcat was using port 8080, preventing me from accessing that port with the Tomcat version downloaded by Maven, resulting in a conflict with the one installed in my OS with APT.

I had to edit my 'pom.xml' to make my project work properly with a more recent version of JDK since the original one was written almost five years ago. So I have it like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.6.3</version> <!-- Actualiza a la última versión estable -->
        <relativePath/>
    </parent>
    <groupId>com.example</groupId>
    <artifactId>final</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <name>Final_v7_7</name>
    <description>Demo project for Spring Boot</description>

    <properties>
        <java.version>11</java.version> <!-- Cambia a JDK 21 -->
    </properties>

    <dependencies>
        <!-- Dependencias de Spring Boot -->
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-thymeleaf</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
        </dependency>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-test</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>log4j</groupId>
            <artifactId>log4j</artifactId>
            <version>1.2.17</version>
        </dependency>

        <!-- Dependencias de Hibernate -->
        <dependency>
            <groupId>org.hibernate</groupId>
            <artifactId>hibernate-core</artifactId>
            <version>5.6.6.Final</version>
        </dependency>
        <dependency>
            <groupId>mysql</groupId>
            <artifactId>mysql-connector-java</artifactId>
            <version>8.0.27</version>
        </dependency>

        <!-- Otras dependencias -->
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>1.7.32</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.1</version>
                <configuration>
                    <source>11</source>
                    <target>11</target>
                    <debug>true</debug>
                </configuration>
            </plugin>

            <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
            </plugin>
            <!--
            	Commented this plugin inclusion, since it brought conflicts. Worked fine after that.
                <plugin>
                    <groupId>org.codehaus.mojo</groupId>
                    <artifactId>exec-maven-plugin</artifactId>
                    <version>3.2.0</version>
                    <executions>
                        <execution>
                            <goals>
                                <goal>exec</goal>
                            </goals>
                        </execution>
                    </executions>
                    <configuration>
                        <executable>mvn</executable>
                         Otros parámetros de configuración, como argumentos del comando, directorio de trabajo, etc. 
                    </configuration>
                </plugin>
            -->
        </plugins>
    </build>
</project>
```

Notice I've specified the Java version to 11. When trying to run it with version 21, I was getting lots of errors.

The "/app" directory, mentioned in the first post was, in fact the content of the folder: "/var/lib/flatpak/app/org.apache.netbeans/current/<large_hex_number>/files". It was actually, a virtual folder created by Flatpak, non-existent in the real filesystem.

So my conclusions are:

- It's not recommended at all to install NetBeans via Flatpak or the OS Software Manager. The best option is to download the ".deb" package from the official webpage and install it. The Flatpak version runs in a "virtual" environment and doesn't recognize the packages installed directly in the OS.

- The version 21 of Apache NetBeans has a few bugs, at least when running on Linux Mint 21.3. Therefore, installing Apache NetBeans 20 is the best choice since it fits very well with a standard installation of this OS.


  [1]: https://i.stack.imgur.com/wPgSH.jpg
  [2]: https://i.stack.imgur.com/uIqXS.jpg
  [3]: https://i.stack.imgur.com/cYho3.jpg

**Reply by neilcsmith-net** [reply]
> `sudo update-alternatives --config java`

This is not going to work with NetBeans without also using `sudo update-alternatives --config javac`.  You should then be fine without setting `JAVA_HOME` or messing around with the deb installed files.

> The version 21 of Apache NetBeans has a few bugs, at least when running on Linux Mint 21.3. 

Please elaborate and/or open issues.

**Reply by leandrocaplan** [reply]
@neilcsmith-net

I guess there should be a better way to achieve my wanted result. I've just has posted here the way it worked for me, after hours of tries and failures, since I've posted this issue a week ago, and I had no reply at all.

About the bug of Apache NetBeans 21, it happened that when I clicked on "Maven" tab in Java configuration, it showed me just an empty rectangle. I would upload a screenshot, but I've uninstalled that version, so then I'm not able to do it. So then, I think I should soon open an issue about that.

Anyway, I appreciate your reply for this issue.
Greetings

Leandro
